**Лабораторная работа №4**  
**Метод опорных векторов**  
Выполнила: Иппо Вера, гр. P4117  

**Цель работы:** получить практические навыки работы с методом дискриминантного анализа и визуализацией данных на практических примерах с использованием языка программирования python.  

**Исходные данные**
Датасет: https://archive.ics.uci.edu/ml/datasets/Letter+Recognition  
Предметная область: буквы латинского алфавиита  
Задача: определить, какой из букв латинского алфавита соответствует набор характеристик ее написания.  
Количество записей: 20000  
Количество атрибутов: 16  

**Атрибуты:**  
1. lettr capital letter (26 values from A to Z)  
2. x-box horizontal position of box (integer)  
3. y-box vertical position of box (integer)  
4. width width of box (integer)  
5. high height of box (integer)  
6. onpix total # on pixels (integer)  
7. x-bar mean x of on pixels in box (integer)  
8. y-bar mean y of on pixels in box (integer)  
9. x2bar mean x variance (integer)  
10. y2bar mean y variance (integer)  
11. xybar mean x y correlation (integer)  
12. x2ybr mean of x * x * y (integer)  
13. xy2br mean of x * y * y (integer)  
14. x-ege mean edge count left to right (integer)  
15. xegvy correlation of x-ege with y (integer)  
16. y-ege mean edge count bottom to top (integer)   
17. yegvx correlation of y-ege with x (integer)  


In [15]:
import pandas
from sklearn.model_selection import train_test_split
from sklearn import svm
from time import time

In [11]:
#загрузка датасета
def load_dataset(filename):
	csv_dataset = pandas.read_csv(filename, header=None).values
	dataset=csv_dataset
	return dataset
	
#разбиение на обучающую и тестовую выборки
def split_dataset(dataset,test_size):
	letter_attr = dataset[:,1:] # список атрибутов (признаков) для каждой буквы
	letter_class = dataset[:,0] # классы букв
	data_train, data_test, class_train, class_test = train_test_split(letter_attr, letter_class, test_size=test_size)
	return data_train, class_train, data_test, class_test

dataset = load_dataset("letter-recognition.csv")
data_train, class_train, data_test, class_test = split_dataset(dataset,0.4)

**Сравнение по ядру**

In [17]:
def calcLinearKernel(data_train, class_train, data_test, class_test):
    # Ядро Linear
    t0 = time()
    linear = svm.SVC(kernel="linear").fit(data_train, class_train)
    print("Затраченное время:", time() - t0, "c.")
    print("Точность ядра Linear: {:.2%}".format(linear.score(data_test, class_test)))

In [20]:
calcLinearKernel(data_train, class_train, data_test, class_test)

Затраченное время: 5.455849647521973 c.
Точность ядра Linear: 84.72%


In [21]:
def calcRBFKernel(data_train, class_train, data_test, class_test):
    # Ядро RBF
    t1 = time()
    rbf = svm.SVC(kernel="rbf").fit(data_train, class_train)
    print("Затраченное время:", time() - t1, "c.")
    print("Точность ядра RBF: {:.2%}".format(rbf.score(data_test, class_test)))

calcRBFKernel(data_train, class_train, data_test, class_test)

Затраченное время: 13.619396924972534 c.
Точность ядра RBF: 96.54%


In [23]:
def calcSigmoldKernel(data_train, class_train, data_test, class_test):
    # Ядро Sigmoid
    t2 = time()
    sigmoid = svm.SVC(kernel="sigmoid").fit(data_train, class_train)
    print("Затраченное время: ", time() - t2, "c.")
    print("Точность ядра Sigmoid: {:.2%}".format(sigmoid.score(data_test, class_test)))

calcSigmoldKernel(data_train, class_train, data_test, class_test)

Затраченное время:  14.770133972167969 c.
Точность ядра Sigmoid: 3.76%


**Сравнение по параметру С**

In [28]:
def calcWhenC(data_train, class_train, data_test, class_test,c):
    linearC = svm.SVC(kernel="linear",C=c).fit(data_train, class_train)
    rbfC = svm.SVC(kernel="rbf",C=c).fit(data_train, class_train)
    sigmoidC = svm.SVC(kernel="sigmoid",C=c).fit(data_train, class_train)
    print("C = ",c)
    print("Точность ядра Linear: {:.2%}".format(linearC.score(data_test, class_test)))
    print("Точность ядра RBF: {:.2%}".format(rbfC.score(data_test, class_test)))
    print("Точность ядра Sigmoid: {:.2%}".format(sigmoidC.score(data_test, class_test)))

In [30]:
calcWhenC(data_train, class_train, data_test, class_test,0.5)

C =  0.5
Точность ядра Linear: 84.91%
Точность ядра RBF: 95.74%
Точность ядра Sigmoid: 3.76%


In [31]:
calcWhenC(data_train, class_train, data_test, class_test,10)

C =  10
Точность ядра Linear: 84.70%
Точность ядра RBF: 97.17%
Точность ядра Sigmoid: 3.76%


In [32]:
calcWhenC(data_train, class_train, data_test, class_test,50)

C =  50
Точность ядра Linear: 84.55%
Точность ядра RBF: 97.16%
Точность ядра Sigmoid: 3.76%


| Ядро  |  С=0.5 |  C=1   | C=10 | C=50 |  
|-------|--------|--------|------|------|  
|Linear | 84.91% | 84.72% |84.70%|84.55%|
|RBF    | 95.74% | 96.54% |97.17%|97.16%|
|Sigmold| 3.76%  | 3.76%  |3.76%|3.76%|

**Сравнение по параметру gamma ядра RBF**  
При варьировании параметра С наилучший результат показало ядро RBF с С=10. Для определения возможности дальнейшего улучшения точности подберем оптимальный параметр gamma.

In [ ]:
def calcWhenGamma(data_train, class_train, data_test, class_test,gamma):
    rbfGamma = svm.SVC(kernel="rbf", C=10, gamma=gamma).fit(data_train, class_train)
    print("Gamma = ",gamma,":\t{:.2%}".format(rbfGamma.score(data_test, class_test)))

In [39]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.1)

Gamma = 0.1:	96.86%


In [40]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.5)

Gamma = 0.1:	73.29%


In [41]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.01)

Gamma = 0.1:	96.36%


In [42]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.05)

Gamma = 0.1:	97.31%


In [44]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.06)

Gamma = 0.1:	97.15%


In [45]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.04)

Gamma = 0.1:	97.41%


In [46]:
calcWhenGamma(data_train, class_train, data_test, class_test,0.03)

Gamma = 0.1:	97.41%


**Вывод**  
В данной лабораторной работе была исследована эффективность применения метода опорных векторов для рассматриваемого датасета. Наилучшие результаты по скорости и приемлемые по точности показало ядро Linear, а ядро Sigmold, как оказалось, не подходит для решении задачи распознавания символов в этом датасете, так как его точность меньше 4%. Наилучший результат по точности показало ядро RBF c параметрами C=10 и gamma = 0.04.